Add gazebo support for IRB1200 5/90 variant

tulku opened the pull request:
* Adds the irb1200_gazebo package with support for the 5 90 variant using effort controllers.
* Adds inertial values to abb_irb1200_support irb1200_5_90 xacro file

This is built upon the initial work of @stwirth (https://github.com/ros-industrial/abb_experimental/pull/98) and @jonbinney

## Intertials

Meshlab was used to calculate the inertia tensor, the center of mass and the volume of each collision mesh. This is the method explained in this [very well known tutorial](http://gazebosim.org/tutorials?tut=inertia). Meshes were scaled 10 times to get better results.

To calculate each link mass a constant density in the entire robot was assumed and then distributed the mass using the calculated volume of each link.

I used the obtained values for all joints except for `joint_6` where they made the robot "crash". Maybe it's because the values are too small?

The new inertials look like this:
![screenshot from 2019-02-28 19-56-24](https://user-images.githubusercontent.com/764126/53590851-fbf65680-3b92-11e9-9ef4-46267a9e9393.png)

I tested the cylinder approximation for link 6 but those values failed in the same way, and were also very small.

## Dynamics and control

Gazebo plotting tool was use to get and idea of the robot responses to motion commands. The original values made the robot oscillate significantly while moving, as can be seen on this plot:
![screenshot from 2019-02-28 18-58-01](https://user-images.githubusercontent.com/764126/53590931-26481400-3b93-11e9-8705-d87d78874003.png)

The new values make the robot move smoothly and pretty quickly:
![screenshot from 2019-02-28 19-43-44](https://user-images.githubusercontent.com/764126/53590304-c2711b80-3b91-11e9-91bc-dba072267589.png)

I'm pretty sure that we can find even better values if we keep tuning, for example to reduce even more the steady state errors, but I believe that the proposed values are already very adequate.

## Testing

Start the simulation:
```
$ roslaunch abb_irb1200_gazebo  irb1200_5_90_gazebo.launch
```
Test the controllers:
```
$ rosrun rqt_joint_trajectory_controller rqt_joint_trajectory_controller arm_controller/command:=joint_path_command arm_controller/state:=feedback_states
```

tulku committed:
Add gazebo support for IRB1200 5/90 variant

* Adds inertial values to abb_irb1200_support irb1200_5_90 xacro file
* Adds the irb1200_gazebo package with support for the 5 90 variant
using effort controllers.

This is built upon the initial work of @stwirth
(https://github.com/ros-industrial/abb_experimental/pull/98) and
@jonbinnie.

gavanderhoorn commented on abb_irb1200_gazebo/package.xml:
```diff
@@ -0,0 +1,35 @@
+    </p>
+    <p>
+      This package contains configuration and launch files
+      to simulate the ABB IRB 1200 with Gazebo.
```
Could you please mention the specific variant for which you added Gazebo support?

gavanderhoorn replied:
Also: I believe it would be good to document *where* the dynamic properties came from. Especially since these are guestimates (however nice they are), it would be good for future users to understand how you arrived at them to be able to gauge their usability in certain (their) contexts.

tulku replied:
I added some information to the `package.xml` files. Let me know if that's not enough.

gavanderhoorn commented on abb_irb1200_gazebo/package.xml:
```diff
@@ -0,0 +1,35 @@
+  <exec_depend>ros_controllers</exec_depend>
+
+  <export>
+  </export>
```
Please remove.

gavanderhoorn commented on abb_irb1200_gazebo/urdf/irb1200_5_90.xacro:
```diff
@@ -0,0 +1,28 @@
+  <xacro:abb_irb1200_5_90 prefix="" />
+
+  <!-- get gazebo add-ons -->
+  <xacro:abb_irb1200_5_90_gazebo_spec prefix="" />
```
The IRB 120 Gazebo pkg includes the Gazebo/`ros_control` specific bits in this file (see [abb_irb120_gazebo/irb120_3_58_macro.xacro](https://github.com/ros-industrial/abb_experimental/blob/f8e2bf903c0875b0799521ec383d15b76344c249/abb_irb120_gazebo/urdf/irb120_3_58_macro.xacro), with the `world` link bits being added by the [top-level xacro](https://github.com/ros-industrial/abb_experimental/blob/f8e2bf903c0875b0799521ec383d15b76344c249/abb_irb120_gazebo/urdf/irb120_3_58.xacro)).

Would it make sense to do that here as well?

Did you have a specific reason for separating it all out into the `abb_irb1200_5_90_gazebo_spec` macro?

tulku replied:
Not really, everything is now in the `_gazebo` macro (removed the `_spec` also, as other robots don't have it)

gavanderhoorn commented on abb_irb1200_gazebo/urdf/irb1200_5_90_macro.xacro:
```diff
@@ -0,0 +1,86 @@
+
+    <!-- Gazebo-specific link properties -->
+    <gazebo reference="${prefix}base_link">
+      <turnGravityOff>true</turnGravityOff>
```
Why is gravity turned off here?

tulku replied:
Gazebo support for the irb1200 was based off the irb120 package which also [disables gravity](https://github.com/ros-industrial/abb_experimental/blob/ce65133979980bab79a4f8f8c3afe78c4eb126c7/abb_irb120_gazebo/urdf/irb120_3_58_macro.xacro#L67-L94) for the arm links.

I now tried to make the model work without those lines (I agree that it would be better not to have them), without success. I needed to increase the friction and in general the forces applied which made the controller sometimes unstable and not perform very well. I would leave it as is.

gavanderhoorn commented on abb_irb1200_support/urdf/irb1200_5_90_macro.xacro:
```diff
@@ -104,42 +145,48 @@
       <parent link="${prefix}base_link"/>
       <child link="${prefix}link_1"/>
-      <limit effort="0" lower="-2.967" upper="2.967" velocity="5.027"/>
+      <limit effort="1000" lower="-2.967" upper="2.967" velocity="5.027"/>
```
If we don't know the effort limits, we usually keep them at `0`. Is that not possible?

tulku replied:
It seems that when using `effort_controllers` the max effort specified in the URDF is used to limit the joint effort. Making it `0`, does not seem to mean that there is no limit, rather, it means that the joint cannot do any effort.

Trying to trace a bit the code to justify this I found [where the limits are loaded](https://github.com/ros-simulation/gazebo_ros_pkgs/blob/2901fe6cc16c86a68cb01d664a45276cac9ec712/gazebo_ros_control/src/default_robot_hw_sim.cpp#L393) and that the limits [are enforced](https://github.com/ros-simulation/gazebo_ros_pkgs/blob/2901fe6cc16c86a68cb01d664a45276cac9ec712/gazebo_ros_control/src/default_robot_hw_sim.cpp#L302-L303) when 'writing' to the sim, which results on [all joint commands being 0](https://github.com/ros-controls/ros_control/blob/f38bf6d7c21e744e37b254d86f1ab7796ea803f7/joint_limits_interface/include/joint_limits_interface/joint_limits_interface.h#L309).

So, to the best of my knowledge, we cannot leave `effort=0` if we want to use effort controllers (instead of position controllers as used [in other packages](https://github.com/ros-industrial/abb_experimental/blob/ce65133979980bab79a4f8f8c3afe78c4eb126c7/abb_irb120_gazebo/config/irb120_3_58_arm_controller.yaml#L2)).

gavanderhoorn commented:
Thanks for the PR @tulku :+1:.

I've left some questions and comments.

gavanderhoorn commented:
@tulku: would it be possible to address the requested changes?

tulku commented:
@gavanderhoorn I need to take care of other things now, but I'd like to do them when I have time.

tulku committed:
Re-arrange abb_irb1200 xacros

tulku committed:
Document how the values were obtained

tulku commented:
@gavanderhoorn I've updated this PR addressing some of your comments. I can rebase on kinetic-devel when its ready to merge.

gavanderhoorn committed:
irb1200: clarify where inertial, dynamics and effort values came from.

gavanderhoorn committed:
irb1200_gazebo: update gazebo pkg manifest.

Changes:

 - add ROS-Industrial maintainers
 - bring description in-line with other packages
 - sort dependencies alphabetically
 - mark package as architecture independent

gavanderhoorn committed:
irb1200_gazebo: don't depend on metapackages.

gavanderhoorn commented:
@tulku: apologies for taking so long, but I've got a few commits I'd like to add to this PR.

Minor changes to the package manifest and `.macro_xacro` only.

I don't believe you've checked "Allow edits from maintainers", so I can't push them directly.

Would you be willing to enable that? I believe you can still click the checkmark. It should be on the right side of the first post in this PR (see also [Allowing changes to a pull request branch created from a fork](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork)).

If not I can open a PR against your PR.

tulku commented:
@gavanderhoorn thanks for looking into this PR!

The "Allow edits from maintainers" was indeed marked. I unmarked and marked it again to see if that helps... If it does not work, I can quickly merge a PR against my branch.

gavanderhoorn commented:
Just tried again and I can't push to `feature/abb_irb1200_5_90_gazebo` on your fork.

Do you have branch protection configured?

tulku commented:
@gavanderhoorn No, I don't see any branch protection configured. I just invited you as collaborator to the fork, with that you should be able to push.

gavanderhoorn reviewed (approved):
Looks good like this.

gavanderhoorn commented:
Thanks again for the PR @tulku :+1: 

performance is indeed quite nice with these gains and dynamics properties.

Thanks for taking the time to contribute it.

gavanderhoorn commented:
Squash merging as all commits but the first one are fixups or review changes.
